### PR TITLE
Add tests and Json parsing

### PR DESCRIPTION
**Pull request by epicallan**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,7 +5,8 @@
   "scripts": {
     "build": "tsc",
     "start": "npm run build && node dist/index.js",
-    "dev": "tsc --watch"
+    "dev": "tsc --watch",
+    "test": "jest"
```
Comment (epicallan) [inline package.json:9]: Comment: If you run the tests command i.e `npm run test` you should see all the current tests passing.

File: src/Json/parser.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import ParseException from "../exceptions/ParseException";
+
+type NonNullableUnknown = NonNullable<unknown>;
```
Comment (epicallan) [inline src/Json/parser.ts:3]: Question: Do you know the difference between the `unknown` type & `any`?

File: src/models/User.ts [hunk 3]
```diff
@@ -1,18 +1,21 @@
...
 
-export type UserJson = Record<string, string>;
+export function decode(json: Json): User {
+  const body = parseJson(json, ["name", "email"]);
 
-export function decode(body: UserJson): User {
   return {
-    name: new Name(body.name),
-    email: new Email(body.email),
+    name: new Name(body.name as string),
```
Comment (epicallan) [inline src/models/User.ts:17]: Question: Can you explain whats happening at this line more specifically why `as strng` is  required?

File: src/tests/models/User/User.test.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            city: "kampala",
+          };
+
+          expect(() => decode(userJson)).toThrow(
+            new ParseException("Json has unexpected keys: city")
+          );
+        });
+      });
+
+      describe("when name is less than 3 characters", () => {
```
Comment (epicallan) [inline src/tests/models/User/User.test.ts:52]: Comment: you will notice I have only added tests for validating `name` but we also add tests for validating the email field in the future.

File: src/Json/parser.ts (hunk 2 above)
Comment (katumwalauren) [reply src/Json/parser.ts:3]: Any allows any type of assignment and can perform code operations without having errors  eg    password: { type: Any, required: true }, if am to have a string or a number, my code doesn't give errors cos it allows the different types  while for unknown, errors are given cos specifications are needed to proceed with a given code

Comment (katumwalauren) [reply src/Json/parser.ts:3]: Any can allow assignment to any type, and you can perform any operation on it without TypeScript giving any errors. eg 
  name: { type: Any, required: true }: so, any type can allow me to use the string or numbers type without errors
While unknown brings errors when coding until one specifies the types to be used

File: src/models/User.ts (hunk 3 above)
Comment (katumwalauren) [reply src/models/User.ts:17]: it is used to tell TypeScript to treat a value as a specific type. The name property is defined twice, which can lead to unexpected behavior. it ensures properties are uniquely defined in an object.

Comment (epicallan) [reply src/models/User.ts:17]: >it is used to tell TypeScript to treat a value as a specific type.  

This is correct. 

> The name property is defined twice

No, it's not. You are just confused by `git diff` output. See this video https://www.youtube.com/watch?v=w6nstXbiL40

File: src/Json/parser.ts (hunk 2 above)
Comment (epicallan) [reply src/Json/parser.ts:3]: Thats correct

**Pull request merged**